`AttributeError`s in `problem.show()` for certain arithmetic expressions

If you add things in a particular order, you can get this confusing error message when generating the (very cool!) visualization of reads and writes. This seems to be triggered when you add a shared read (tracked as a `Scalar`) to a mathematical expression (tracked as a `ScalarHistory`).

Somewhat minimal repro:

```python
TPB = 8
def error_repro(cuda):
    def call(out, a, size) -> None:
        shared = cuda.shared.array(TPB, numba.float32)
        i = cuda.blockIdx.x * cuda.blockDim.x + cuda.threadIdx.x
        local_i = cuda.threadIdx.x

        val = a[i] * a[i]
        val += 1.0
        shared[local_i] = 0.0
        cuda.syncthreads()
        shared[local_i] = shared[local_i] + val  # <- bad line!
        # shared[local_i] = val + shared[local_i]  # <- this works
        cuda.syncthreads()
        out[i] = shared[local_i]

    return call


SIZE = 8
out = np.zeros(SIZE)
a = np.arange(SIZE)
problem = CudaProblem(
    "Pooling",
    error_repro,
    [a],
    out,
    [SIZE],
    threadsperblock=Coord(TPB, 1),
    blockspergrid=Coord(1, 1),
    spec=pool_spec,
)
problem.show()
```

Traceback:
```
AttributeError                            Traceback (most recent call last)
<ipython-input-169-389d915eeb33> in <cell line: 32>()
     30     spec=pool_spec,
     31 )
---> 32 problem.show()

1 frames
/content/lib.py in show(self, sparse)
    389     def show(self, sparse=False):
    390         results = self.run_python()
--> 391         self.score(results)
    392         return draw_results(results, self.name,
    393                             self.threadsperblock.x, self.threadsperblock.y, sparse)

/content/lib.py in score(self, results)
    373                         count["shared_writes"] += 1
    374                     for ins in inc[1].inputs:
--> 375                         if ins.location[0].startswith("S"):
    376                             count["shared_reads"] += 1
    377                         else:

AttributeError: 'ScalarHistory' object has no attribute 'location'
```

Another confusing error message is this one, where I'm already doing the thing that the error suggests:

```python
TPB = 8
def error_repro(cuda):
    def call(out, a, size) -> None:
        shared = cuda.shared.array(TPB, numba.float32)
        i = cuda.blockIdx.x * cuda.blockDim.x + cuda.threadIdx.x
        local_i = cuda.threadIdx.x

        val = a[i]
        val += 1
        out[i] = val

    return call


SIZE = 8
out = np.zeros(SIZE)
a = np.arange(SIZE)
problem = CudaProblem(
    "Pooling",
    error_repro,
    [a],
    out,
    [SIZE],
    threadsperblock=Coord(TPB, 1),
    blockspergrid=Coord(1, 1),
    spec=pool_spec,
)
problem.show()
```

```
AssertionError                            Traceback (most recent call last)
<ipython-input-173-8b13254f219a> in <cell line: 28>()
     26     spec=pool_spec,
     27 )
---> 28 problem.show()

3 frames
/content/lib.py in show(self, sparse)
    388 
    389     def show(self, sparse=False):
--> 390         results = self.run_python()
    391         self.score(results)
    392         return draw_results(results, self.name,

/content/lib.py in run_python(self)
    354 
    355                 c = Cuda(block, self.threadsperblock, pos)
--> 356                 fn(c)(out, *a, *self.args)
    357                 c.finish()
    358                 results[block][pos] =  (tt, a, c, out)

/content/lib.py in call(out, a, size)
      7 
      8         val = a[i]
----> 9         val += 1
     10         out[i] = val
     11 

/content/lib.py in __iadd__(self, other)
     46 
     47     def __iadd__(self, other):
---> 48         assert False, "Instead of `out[] +=` use a local variable `acc + =`"
     49 
     50 class Table:

AssertionError: Instead of `out[] +=` use a local variable `acc + =`
```

--- a/lib.py
+++ b/lib.py
@@ -23,8 +23,9 @@ class ScalarHistory:
             return self
         if isinstance(b, Scalar):
             return ScalarHistory(self.last_fn, self.inputs + [b])
-
-        return ScalarHistory(self.last_fn, self.inputs + b.inputs)
+        if isinstance(b, ScalarHistory):
+            return ScalarHistory(self.last_fn, self.inputs + b.inputs)
+        return NotImplemented
         
 class Scalar:
     def __init__(self, location):
@@ -33,8 +34,9 @@ class Scalar:
     def __mul__(self, b):
         if isinstance(b, (float, int)):
             return ScalarHistory("id", [self])
-        return ScalarHistory("*", [self, b])
-
+        if isinstance(b, Scalar):
+            return ScalarHistory("*", [self, b])
+        return NotImplemented
 
     def __radd__(self, b):
         return self + b
@@ -42,10 +44,11 @@ class Scalar:
     def __add__(self, b):
         if isinstance(b, (float, int)):
             return ScalarHistory("id", [self])
-        return ScalarHistory("+", [self, b])
-
-    def __iadd__(self, other):
-        assert False, "Instead of `out[] +=` use a local variable `acc + =`"
+        if isinstance(b, Scalar):
+            return ScalarHistory("+", [self, b])
+        if isinstance(b, ScalarHistory):
+            return ScalarHistory("+", [self] + b.inputs)
+        return NotImplemented
     
 class Table:
     def __init__(self, name, array):
@@ -74,8 +77,9 @@ class Table:
             assert False, "bad size"
         if isinstance(val, Scalar):
             val = ScalarHistory("id", [val])
-        if isinstance(val, float):
+        if isinstance(val, (float, int)):
             return
+        assert isinstance(val, ScalarHistory), "Assigning an unrecognized value"
         self.incoming.append((index, val))
 
 @dataclass(frozen=True, eq=True)

